flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPool.java

```diff
@@ -0,0 +1,353 @@
+import org.apache.flink.core.memory.MemorySegmentFactory;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.TimeoutException;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * A fixed-size {@link MemorySegment} pool used by batch shuffle for shuffle data read (currently
+ * only used by sort-merge blocking shuffle).
+ */
+@Internal
+public class BatchShuffleReadBufferPool {
+
+    private static final Logger LOG = LoggerFactory.getLogger(BatchShuffleReadBufferPool.class);
+
+    /** Minimum total memory size in bytes of this buffer pool. */
+    public static final int MIN_TOTAL_BYTES = 32 * 1024 * 1024;
+
+    /**
+     * Memory size in bytes can be allocated from this buffer pool for a single request (8M is for
+     * better sequential read).
+     */
+    public static final int NUM_BYTES_PER_REQUEST = 8 * 1024 * 1024;
+
+    /** Total direct memory size in bytes can can be allocated and used by this buffer pool. */
+    private final long totalBytes;
+
+    /**
+     * Maximum time to wait in milliseconds when requesting read buffers from this buffer pool
+     * before throwing an exception.
+     */
+    private final long requestTimeout;
+
+    /** The number of total buffers in this buffer pool. */
+    private final int numTotalBuffers;
+
+    /** Size of each buffer in bytes in this buffer pool. */
+    private final int bufferSize;
+
+    /** The number of buffers to be returned for a single request. */
+    private final int numBuffersPerRequest;
+
+    /**
+     * The maximum number of buffers can be allocated from this buffer pool for a single buffer
+     * requester.
+     */
+    private final int maxBuffersPerRequester;
+
+    /** All available buffers in this buffer pool currently. */
+    @GuardedBy("buffers")
+    private final Queue<MemorySegment> buffers = new ArrayDeque<>();
+
+    /** Account for all the buffers requested per requester. */
+    @GuardedBy("buffers")
+    private final Map<Object, Counter> numBuffersAllocated = new HashMap<>();
+
+    /** Whether this buffer pool has been destroyed or not. */
+    @GuardedBy("buffers")
+    private boolean destroyed;
+
+    /** Whether this buffer pool has been initialized or not. */
+    @GuardedBy("buffers")
+    private boolean initialized;
+
+    public BatchShuffleReadBufferPool(long totalBytes, int bufferSize) {
+        // 5 min default buffer request timeout
+        this(totalBytes, bufferSize, 5 * 60 * 1000);
+    }
+
+    public BatchShuffleReadBufferPool(long totalBytes, int bufferSize, long requestTimeout) {
```

Comment:
You are right, it only for test currently, because the default 5min timeout is a little long for test. BatchShuffleReadBufferPoolTest#testRequestBuffersTimeout and BatchShuffleReadBufferPoolTest#testNoBufferTimeoutIfHoldingBuffers use it.

Reply:
Let's annotate it as `@VisibleForTesting` and make it package private.